docs/architecture/adr-046-light-client-implementation.md

```diff
@@ -127,6 +128,21 @@ cases of adjacent and non-adjacent headers. In the former case, it compares the
 hashes directly (2/3+ signed transition). Otherwise, it verifies 1/3+
 (`trustLevel`) of trusted validators are still present in new validators.
 
+### Bisection algorithm details
+
+Non-recursive bisection algorithm was implemented despite the spec containing
+the recursive version. There are two major reasons:
+
+1) Constant memory consumption => no risk of getting OOM (Out-Of-Memory) exceptions;
+2) Faster finality (see Fig. 1).
```

Comment:
while this is true for the example you show in Fig.1, not sure this is always the case. For example, in another case you can have something like this:

- start with th=1, ih=256
- the +1/3rd check fails a few times and interim headers are retrieved and checked, let's say: [1], 8, 16, 32, 64, 128, [256]
- 8 verifies against 1 so now we have th=8, ih=256
- again the +1/3rd check fails and other headers are retrieved: [8], 16, 24, 40, 72, 136, [256]
- 16 verifies against 8 and at this point th=16, ih=256
-...

The downloaded headers are not kept around (so we save on memory) but more headers and vals need to be retrieved.
Was the OOM something you saw? If we do the math for th and ih being 3wk apart (trusting period) and 1block/sec we probably have a ~20 depth in recursion with ~20blocks retrieved.

Reply:
> The downloaded headers are not kept around (so we save on memory) but more headers and vals need to be retrieved.

Hmm.. You're right! Looks like we want to save those interim headers.

Thanks!

> Was the OOM something you saw? 

no, theoretical concern

Reply:
I am not too sure if caching the interim headers will work, because the range changes so therefore the headers to be requested by the provider will be all different.

Reply:
Can you give an example?

Won't we continue with same headers [8], 16, 24, 40, 72, 136, [256]?

Reply:
No I thought we would find the halfway point between 8 and 256

Reply:
it also defeats a bit the purpose of the current non-recursive version. You can still do a non-recursive version by simulating the recursive stack and just store the headers (instead of full stack frames) and keep the same range.  
But maybe before you make more changes, it would be interesting to get some header/vals data from mainnet, and run both the version you have and the old recursive one to see differences in memory usage and performance. Or just use generated data.

Reply:
yes I would be interested in checking that out

Reply:
created https://github.com/tendermint/tendermint/issues/4504

Reply:
Also, it may be likely that a +1/3rd power change is a rare event in the 3-wk sliding window worth of blocks and the current non-recursive version may turn out to be the best. But again, this is something worth investigating.